### ISSUE TITLE
log_softmax(x) != x - logsumexp(x)

### ISSUE THREAD

**Issue by dniku** [pytorch/pytorch#56340]
## 🐛 Bug

There is a rather large numerical difference between `log_softmax` as computed by `torch.nn.functional.log_softmax` and `x - logsumexp(x)`.

```python
import torch

def log_softmax(x, dim=-1):
    return x - torch.logsumexp(x, dim=dim, keepdim=True)

a = torch.tensor([[800, 800, 400]], dtype=torch.float32)
pt = torch.nn.functional.log_softmax(a, dim=-1)
print(pt)
# tensor([[  -0.6931,   -0.6931, -400.6931]])

my = log_softmax(a, dim=-1)
print(my)
# tensor([[  -0.6932,   -0.6932, -400.6932]])

print(torch.abs(pt - my))
# tensor([[2.9087e-05, 2.9087e-05, 3.0518e-05]])

print(torch.allclose(pt, my))
# False
```

## To Reproduce

https://colab.research.google.com/drive/1B1exQKAC_sMBMqiMzY42gw97YYq1PVvP?usp=sharing

## Expected behavior

`torch.allclose` should pass.

## Environment

Colab.

```
PyTorch version: 1.8.1+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: 11.0.221
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.1+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.9.1
[pip3] torchvision==0.9.1+cu101
[conda] Could not collect
```

**Comment by vishwakftw** [comment]
You are missing one step while computing the softmax which is the max-trick. This is done to improve precision and is given by `softmax(x) = softmax(x - max(x))`.

So ideally your comparison should be with `x - max(x, dim=-1) - torch.logsumexp(x - max(x, dim=-1), dim=-1)`.

**Comment by dniku** [comment]
Strictly speaking, this isn't necessary for numerical stability when I'm using `logsumexp`, since `logsumexp` never blows up on any inputs.

I have also verified that `log_softmax()` implemented as

```python
def log_softmax(x, dim=-1):
    tmp = x - x.max(dim=dim, keepdims=True)[0]
    return tmp - torch.logsumexp(tmp, dim=dim, keepdim=True)
```

passes the `allclose()` check.

I'm not sure if this issue should be kept open. Please close it or keep it open as you see fit.

**Comment by abcde2342** [comment]
I think your manual log_softmax is not numerically stable. The right one is like:

```
def log1p_taylor_tensor(s, terms=8):
    total = torch.zeros((), dtype=s.dtype, device=s.device)
    power = torch.ones((), dtype=s.dtype, device=s.device)
    for n in range(1, terms + 1):
        power *= s
        sign = 1.0 if n % 2 == 1 else -1.0
        total = total + sign * power / n
    return total

def log_softmax_taylor_equalmax_component(x, index=0, terms=8):
    a = torch.tensor(np.asarray(x, dtype=np.float32))
    m = torch.max(a)
    shifted = a - m
    max_mask = shifted == 0
    k = torch.sum(max_mask).to(dtype=a.dtype)
    tail = torch.sum(torch.exp(shifted[~max_mask]))
    z = tail / k
    return float((shifted[index] - torch.log(k) - log1p_taylor_tensor(z, terms=terms)).item())
```
